### ISSUE TITLE
Deadlock found when trying to get lock; try restarting transaction

### ISSUE THREAD

**Issue by thomas-advantitge** [vendure-ecommerce/vendure#527]
**Describe the bug**
Deadlock occurs when multiple `createProductVariants` calls are made at the same time.
```
Error: ER_LOCK_DEADLOCK: Deadlock found when trying to get lock; try restarting transaction: 
```
```json
{"response":{"errors":
[{"message":"ER_LOCK_DEADLOCK: Deadlock found when trying to get lock; try restarting transaction","locations":
[{"line":3,"column":9}],"path":["createProductVariants"]}],"data":null,"status":200},"request":{"query":"\n    mutation 
CreateProductVariants($input: [CreateProductVariantInput!]!) {\n        createProductVariants(input: $input) {\n            id\n            sku\n        }\n    }\n","variables":{"input":[{"assetIds":["6374"],"featuredAssetId":"6374","sku":"010168240001","translations":
[{"languageCode":"fr","name":"CORONA EXTRA ONE WAY 35.5CL","customFields":{"content":""}},
{"languageCode":"nl","name":"CORONA EXTRA ONE WAY 35.5CL","customFields":{"content":""}}],"customFields":
{"pieces":1,"totalContent":0.36},"productId":"525","price":100000,"trackInventory":false,"taxCategoryId":"1","optionIds":
["776"]},{"assetIds":["6375"],"featuredAssetId":"6375","sku":"010168240006","translations":
[{"languageCode":"fr","name":"CORONA EXTRA ONE WAY 6X35.5CL","customFields":{"content":""}},
{"languageCode":"nl","name":"CORONA EXTRA ONE WAY 6X35.5CL","customFields":{"content":""}}],"customFields":
{"pieces":6,"totalContent":2.13},"productId":"525","price":100000,"trackInventory":false,"taxCategoryId":"1","optionIds":
["777"]}]}}}
```

**Expected behavior**

**Environment (please complete the following information):**
 - @vendure/core version: 0.16.1
 - Nodejs version: 12.18.3
 - Database (mysql/postgres etc): mysql

**Additional context**
Possible related issue: #242

**Comment by michaelbromley** [comment]
Thanks for the report. This does indeed look to be a problem with the Transaction support. How many concurrent calls to `createProductVariants` is required to trigger this error?

**Comment by thomas-advantitge** [comment]
> Thanks for the report. This does indeed look to be a problem with the Transaction support. How many concurrent calls to `createProductVariants` is required to trigger this error?

Hi @michaelbromley, occurred frequently when having 4 concurrent calls, although it also occurred sometimes executing just 2 parallel calls.

**Comment by michaelbromley** [comment]
Got it. What is the use-case for having so many concurrent calls to `createProductVariants` by the way? Are you doing an automated import of product data?

**Comment by thomas-advantitge** [comment]
We're doing an automated import of products. Some background:
* The first importer version was using the fast importer. However we have a quite complex integration with an external system,  requiring import of categories, stores, prices,...
* The second version was running as a plugin injecting Vendure services. This worked, although we sometimes missed some functionality which was handled in Vendure resolvers. 
* In our current version we're using the GraphQL interface to insert data. Important note here is that an import request from the external system will contain 1 product, collection, channel,... to be created or updated. In this test phase however, we're importing a large dump of data through a script which calls the endpoints in a loop, in order to be able to simulate the same behaviour as if the external system would send data. This however takes a lot of time, especially when uploading assets as well. Therefore, we tried inserting the data in parallel (2 or 4 batches), causing the deadlocks. For the import script, it was easily be solved by just being patient and running sequentially. However, as we're just using the API, this will most likely also happen in production occasionally.

**Comment by michaelbromley** [comment]
I created a new e2e test suite which makes parallel calls to operations designed to induce deadlock ([example](https://github.com/vendure-ecommerce/vendure/blob/3b60bcbe72904342a747e16637bc4d5126f42564/packages/core/e2e/parallel-transactions.e2e-spec.ts#L53)).

I then put in place retry logic in the code which wraps resolvers in a transaction, which seems to solve the issue (at least the e2e tests). This will be available in 0.16.3 which will be released in the next day or so.
